File: 1501~2000/1658. Minimum Operations to Reduce X to Zero/main.py
"""
Source: https://rb.gy/rpd7y
Date: 2023/7/26
Skill:
Ref:
Runtime: TLE
Memory Usage:
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        sz, total = len(nums), sum(nums)
        if total < x:
            return -1
        if total == x:
            return sz
        dp = [[0 for _ in range(sz)] for _ in range(sz)]

        def find_res(left, right, target):
            if left > right:
                return -1
            if dp[left][right] != 0:
                return dp[left][right]
            if nums[left] > target and nums[right] > target:
                dp[left][right] = -1
                return -1
            if nums[left] == target or nums[right] == target:
                dp[left][right] = 1
                return 1
            step = sys.maxsize
            if nums[left] < target:
                step1 = find_res(left + 1, right, target - nums[left])
                if step1 != -1:
                    step = min(step, step1 + 1)
            if nums[right] < target:
                step2 = find_res(left, right - 1, target - nums[right])
                if step2 != -1:
                    step = min(step, step2 + 1)
            dp[left][right] = step if step < sys.maxsize else -1
            return dp[left][right]

        return find_res(0, sz - 1, x)


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1658. Minimum Operations to Reduce X to Zero/main.cpp
/**
 * Source: t.ly/GOihx
 * Date: 2023/9/20
 * Skill:
 * Runtime: 301 ms, faster than 25.32% of C++ online submissions
 * Memory Usage: 151.92 MB, less than 22.91% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        unordered_map<int, int> sum2idx;
        sum2idx[0] = -1;
        int res = INT32_MAX, sz = nums.size(), cur_sum = 0;
        int target = accumulate(begin(nums), end(nums), -x);
        if (target == 0)
            return sz;
        if (target < 0)
            return -1;
        for (int i = 0; i < sz; i++) {
            cur_sum += nums[i];
            if (sum2idx.find(cur_sum - target) != sum2idx.end())
                res = min(res, sz - (i - sum2idx[cur_sum - target]));
            sum2idx[cur_sum] = i;
        }
        return res != INT32_MAX ? res : -1;
    }
};

int main() {

}